Add AEMET OpenData integration

Noltari opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
AEMET stands for Agencia Estatal de Meteorología, which is the Spanish Meteorological Agency.
More info: https://opendata.aemet.es/


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16173

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Noltari committed:
Add AEMET OpenData integration

AEMET stands for Agencia Estatal de Meteorología, which is the Spanish
Meteorological Agency.
More info: https://opendata.aemet.es/

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: fix const declaration

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: station timestamp is UTC

The rest of the timestamps seem to be local.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: town timestamp is UTC

This is not specified in the metadata provided by AEMET, as opposed to station
timestamps (which is specified in the metadata), but it has been deduced from
my own observations.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: tests: fix utcnow value

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: make float/int conversions safe

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: tests: fix config_flow entry state

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: tests: align weather with config_flow and sensors

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
Revert "aemet: town timestamp is UTC"

This reverts commit 8cd67ff5fafb74ae6f17a1af1f85789060b1438e.
Turns out it's not UTC, but local.

Noltari committed:
aemet: add strings

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: update AEMET-OpenData to latest version

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented:
Rebased, since there were `requirements_all.txt` conflicts.

elupus reviewed (commented):
At first glance this looks quite good. Will likely need somebody else's eyes on it but in general good.

elupus commented on homeassistant/components/aemet/__init__.py:
```diff
@@ -0,0 +1,101 @@
+    await weather_coordinator.async_refresh()
+
+    if not weather_coordinator.last_update_success:
+        raise ConfigEntryNotReady
```
Is this needed? is it not possible to let the entry start and just let them end up being unavailable based on coordinator being unavailable?

Noltari replied:
This is actually taken from OpenWeatherMap (https://github.com/home-assistant/core/blob/6b340415b2b6b3585d835f6e17aec7914fa29cd2/homeassistant/components/openweathermap/__init__.py#L59), but I will look into it.

Noltari replied:
@elupus This should be solved with https://github.com/home-assistant/core/pull/45074/commits/5cce58b1460e5b1b9826914f0c59e92d140d0127

elupus replied:
Sort of yes. But now you will keep on logging an error on each failed request. Filling the log (also you can remove some checks for the return value of town == None).

You should either catch your new exception and rethrow a UpdateFailed exception. Or base the exception on UpdateFailed. That way the update coordinator will only log once on failure. See: https://github.com/home-assistant/core/blob/6f4cb18fa88d357eb873101b25159c989c0f9125/homeassistant/helpers/update_coordinator.py#L172

Noltari replied:
@elupus done in https://github.com/home-assistant/core/pull/45074/commits/c96ed23bd03228ca5cb5176ce84d33e70a176442

elupus commented on homeassistant/components/aemet/abstract_aemet_sensor.py:
```diff
@@ -0,0 +1,77 @@
+        unique_id,
+        sensor_type,
+        sensor_configuration,
+        coordinator: DataUpdateCoordinator,
```
You might as well make this a weather_coordinator since you have your own baseclass.

Noltari replied:
This is also based on OpenWeatherMap, but I'll look into it:
https://github.com/home-assistant/core/blob/20f45f8ab9e744de7488fe66c5e7f8aca9b60e2d/homeassistant/components/openweathermap/abstract_owm_sensor.py#L18

elupus commented on homeassistant/components/aemet/weather_update_coordinator.py:
```diff
@@ -0,0 +1,655 @@
+                self._weather_data = data
+        except asyncio.TimeoutError:
+            data = self._weather_data
+            _LOGGER.error("AEMET OpenData API timeout")
```
Just let the timeout error propagate here. The standard handling is to log an error and set last_update_success = False.

elupus commented on homeassistant/components/aemet/weather_update_coordinator.py:
```diff
@@ -0,0 +1,655 @@
+    def _get_daily_forecast_from_weather_response(self, weather_response):
+        if weather_response.daily:
+            parse = False
+            now = dt_util.now()
```
This is a minor nit. I would pass in the timestamp to avoid fetching it more than once. Makes it a bit more deterministic.

elupus commented on homeassistant/components/aemet/weather_update_coordinator.py:
```diff
@@ -0,0 +1,655 @@
+    def _get_hourly_forecast_from_weather_response(self, weather_response):
+        if weather_response.hourly:
+            parse = False
+            now = dt_util.now()
```
As above

elupus commented on .coveragerc:
```diff
@@ -23,6 +23,11 @@ omit =
     homeassistant/components/ads/*
+    homeassistant/components/aemet/abstract_aemet_sensor.py
+    homeassistant/components/aemet/config_flow.py
+    homeassistant/components/aemet/sensor.py
```
Do you need to add senor.py here. You do seem to have tests for it.

Noltari replied:
If I remember correctly there were code coverage errors when I didn't add it.

elupus reviewed (approved):
I only have minor nits. I think this looks good.

Noltari committed:
aemet: remove temperature_feeling from station data

It seems that "tpr" stands for dew point, so this replacemente is incorrect.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: get current time once

Call dt_util.now() once and pass the value to the rest of the functions.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: let TimeoutError propagate to DataUpdateCoordinator

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: log error if condition is not found

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: abstract_aemet_sensor: switch to WeatherUpdateCoordinator

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: remove tested files from .coveragerc

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented:
@elupus all the NITs should be solved now :)

elupus reviewed (approved):
Looks good. We'll see if i dare to push merge. Wouldn't mind some more senior other eyes on it :)

Noltari committed:
aemet: raise TimeNotFound exception

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: base TownNotFound on UpdateFailed exception

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: remove config_flow.py from .coveragerc

There are tests for this file.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: test_config_flow: achieve 100% coverage

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented:
@elupus I've added more tests to reach 100% coverage on config_flow, since I've seen @bdraco mention in other PRs that this is mandatory.

MartinHjelmare commented on homeassistant/components/aemet/__init__.py:
```diff
@@ -0,0 +1,97 @@
+
+    await weather_coordinator.async_refresh()
+
+    hass.data.setdefault(DOMAIN, {})
```
This is already done above.

Noltari replied:
As I already told @MatthewFlamm, I based my integration on OpenWeatherMap, and this is inherited from there: https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/homeassistant/components/openweathermap/__init__.py#L62

MartinHjelmare replied:
Well we shouldn't add not needed code.

MartinHjelmare commented on homeassistant/components/aemet/__init__.py:
```diff
@@ -0,0 +1,97 @@
+
+
+def _filter_domain_configs(elements, domain):
+    return list(filter(lambda elem: elem["platform"] == domain, elements))
```
Please use a list comprehension instead of filter + lambda.

Noltari replied:
As above, this is inherited from OpenWeatherMap: https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/homeassistant/components/openweathermap/__init__.py#L125

MatthewFlamm commented on homeassistant/components/aemet/weather.py:
```diff
@@ -0,0 +1,120 @@
+    async_add_entities([aemet_weather], False)
+
+
+class AemetWeather(WeatherEntity):
```
Can you inherit from CoordinatorEntity, which might clean up this code a bit?

See https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/homeassistant/helpers/update_coordinator.py#L233

Noltari replied:
This is also inherited from OpenWeatherMap: https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/homeassistant/components/openweathermap/weather.py#L33

MartinHjelmare commented on homeassistant/components/aemet/abstract_aemet_sensor.py:
```diff
@@ -0,0 +1,77 @@
+from .weather_update_coordinator import WeatherUpdateCoordinator
+
+
+class AbstractAemetSensor(Entity):
```
Why aren't we inheriting `CoordinatorEntity`?

Noltari replied:
As above, this is inherited from OpenWeatherMap: https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/homeassistant/components/openweathermap/abstract_owm_sensor.py#L9

MartinHjelmare commented on homeassistant/components/aemet/config_flow.py:
```diff
@@ -0,0 +1,104 @@
+                vol.Optional(
+                    CONF_LONGITUDE, default=self.hass.config.longitude
+                ): cv.longitude,
+                vol.Optional(CONF_MODE, default=DEFAULT_FORECAST_MODE): vol.In(
```
Mode should only be part of the options flow. Optional items not related specifically with the connection go there.

MartinHjelmare replied:
But preferably we shouldn't have an option at all. See Matthew's review below.

Noltari replied:
As above, this is inherited from OpenWeatherMap: https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/homeassistant/components/openweathermap/config_flow.py#L78

MartinHjelmare replied:
If there's no issue, please remove the option and let the user enable/disable the entities instead. Start with just the default mode entity enabled by default.

Noltari replied:
I'll mark this as resolved and deal with it in @MatthewFlamm's review.

MartinHjelmare commented on homeassistant/components/aemet/sensor.py:
```diff
@@ -0,0 +1,99 @@
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._weather_coordinator.data.get(self._sensor_type, None)
```
`None` is the default value returned if the key is missing in the dict.
```suggestion
        return self._weather_coordinator.data.get(self._sensor_type)
```

Noltari replied:
As above, this is inherited from OpenWeatherMap: https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/homeassistant/components/openweathermap/sensor.py#L73

MartinHjelmare commented on homeassistant/components/aemet/sensor.py:
```diff
@@ -0,0 +1,99 @@
+    def state(self):
+        """Return the state of the device."""
+        forecasts = self._weather_coordinator.data.get(ATTR_API_FORECAST)
+        if forecasts is not None and len(forecasts) > 0:
```
```suggestion
        if forecasts:
```

Noltari replied:
As above, this is inherited from OpenWeatherMap: https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/homeassistant/components/openweathermap/sensor.py#L97

MartinHjelmare commented on homeassistant/components/aemet/weather_update_coordinator.py:
```diff
@@ -0,0 +1,652 @@
+class WeatherUpdateCoordinator(DataUpdateCoordinator):
+    """Weather data update coordinator."""
+
+    def __init__(self, aemet, latitude, longitude, forecast_mode, hass):
```
Always start with `hass` in the signature if `hass` is present in the signature.

Noltari replied:
As above, this is inherited from OpenWeatherMap: https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/homeassistant/components/openweathermap/weather_update_coordinator.py#L54

MartinHjelmare commented on homeassistant/components/aemet/weather_update_coordinator.py:
```diff
@@ -0,0 +1,652 @@
+        data = {}
+        with async_timeout.timeout(120):
+            weather_response = await self._get_aemet_weather()
+            data = self._convert_weather_response(weather_response)
```
Please only wrap the line that can raise in the try... except block.

Noltari replied:
As above, this is inherited from OpenWeatherMap: https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/homeassistant/components/openweathermap/weather_update_coordinator.py#L73

MatthewFlamm commented on homeassistant/components/aemet/weather_update_coordinator.py:
```diff
@@ -0,0 +1,652 @@
+        return None
+
+    @staticmethod
+    def _get_humidity(day_data, hour):
```
This class is large with a lot of repetition.  Does it make sense to have a single `get_value` or maybe have a few cases like `get_numeric_value`?  Why is the upstream library not supplying numeric values for numeric data? This seems more ideal, and would eliminate the majority need to separate the numeric data from non-numeric data entirely.  We wouldn't need to check in the home assistant code in this case and only a handful of methods would be required.

Noltari replied:
The upstream library is not supplying numeric values in numeric data because the API is not doing it, and the upstream library was implemented with the minimum effort to get this working.

MatthewFlamm replied:
This is more of a style comment, so less critical.  The handling here makes the resulting platform code simpler too.  As you look into splitting into multiple coordinators consider ways to simplify is all.

Noltari replied:
Ok, I'll mark this as resolved and try to simplify it when implementing multiple coordinators...
The problem is that I've got no idea on how to do it right now...

MatthewFlamm replied:
As I said this was more style feedback, and probably differs based on personal preference.  I think it is fine as is.

MartinHjelmare commented on homeassistant/components/aemet/weather_update_coordinator.py:
```diff
@@ -0,0 +1,652 @@
+        return None
+
+
+class AemetWeather:
```
This could be a dataclasses dataclass.

https://docs.python.org/3/library/dataclasses.html

Noltari replied:
As above, this is inherited from OpenWeatherMap:
https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/homeassistant/components/openweathermap/weather_update_coordinator.py#L215

MartinHjelmare commented on tests/components/aemet/test_config_flow.py:
```diff
@@ -0,0 +1,148 @@
+        entry = conf_entries[0]
+        assert entry.state == ENTRY_STATE_LOADED
+
+        await hass.config_entries.async_unload(conf_entries[0].entry_id)
```
Testing setting up and unloading the config entry should be part of tests for the init module, ie not part of the config flow tests.

Noltari replied:
As above, this is inherited from OpenWeatherMap: https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/tests/components/openweathermap/test_config_flow.py#L62

Noltari replied:
@MartinHjelmare honestly, I have no idea on how you want me to do this.
Can you give me an example of an integration that does it the way you want it or any other kind of help?

MartinHjelmare replied:
Here's an example from nws:
https://github.com/home-assistant/core/blob/479ff92acbc9ec4db09b7a300d2beb67f00ab746/tests/components/nws/test_init.py#L10-L37

It's good except for the hass.data checks that we should avoid.

MatthewFlamm commented on homeassistant/components/aemet/weather_update_coordinator.py:
```diff
@@ -0,0 +1,652 @@
+
+        daily = None
+        if self._forecast_mode == FORECAST_MODE_DAILY:
+            daily = self._aemet.get_specific_forecast_town_daily(
```
This class is very convoluted because you have three different api calls.  Can these be separated into separate UpdateCoordinator's?  This would be ideal in the case that you have multiple entity types that require data from different api calls, which seems to be the case.  Otherwise, a failure in one api call results in unavailable data for everything.

Noltari replied:
I need to investigate this.

Noltari replied:
@MatthewFlamm I'm a bit of a newbie here, so I'll need some help to achieve this...
Can you provide examples of other integrations which have multiple update coordinators?
With the other changes requested, AbstractAemetSensor and AemetWeather inherit from CoordinatorEntity, thus containing a single coordinator, so how can I add multiple coordinators now?

Noltari replied:
@MartinHjelmare @MatthewFlamm I'm a bit of a newbie here, so I'll need some help to achieve this...
Can you provide examples of other integrations which have multiple update coordinators?
With the other changes requested, AbstractAemetSensor and AemetWeather inherit from CoordinatorEntity, thus containing a single coordinator, so how can I add multiple coordinators now?

MartinHjelmare replied:
NWS is one example:
https://github.com/home-assistant/core/blob/a67b598971eb5142751dd6642697257b82aa31aa/homeassistant/components/nws/__init__.py#L112-L146

MartinHjelmare replied:
Each entity will only be connected to one coordinator.

MartinHjelmare replied:
Although NWS actually has two coordinators per entity.

Noltari replied:
> Although NWS actually has two coordinators per entity.

Am I allowed to implement multiple coordinators per entity then?

MartinHjelmare replied:
I think the suggestion here is to make one coordinator per forecast mode and this will map 1:1 to the different weather entities which each has a different forecast mode.

Noltari replied:
> I think the suggestion here is to make one coordinator per forecast mode and this will map 1:1 to the different weather entities which each has a different forecast mode.

@MartinHjelmare indeed, that would work for forecast mode, but there are some problems with that:
- Station data replaces predicted hourly data, so we would need on coordinator for hourly forecast and another one for station.
- Sensor/Weather entities created for daily mode won't have precise values, since daily data si split into 6h intervals:
```
"temperatura" : {
        "maxima" : 2,
        "minima" : -1,
        "dato" : [ {
          "value" : -1,
          "hora" : 6
        }, {
          "value" : 0,
          "hora" : 12
        }, {
          "value" : 1,
          "hora" : 18
        }, {
          "value" : 1,
          "hora" : 24
        } ]
      },
```
- So, in the end, for the current temperature, humidity, etc. values on daily weather, we need to rely on the hourly coordinator.

Therefore, the daily mode would need 3 coordinators: `daily, station and hourly`, and the hourly mode would need 2 coordinators: `hourly and station`.

MartinHjelmare replied:
Yeah, I haven't looked into details if this would make things more readable or not. My suggestion is that we save this work for a potential follow up PR.

Noltari replied:
> Yeah, I haven't looked into details if this would make things more readable or not. My suggestion is that we save this work for a potential follow up PR.

Ok, so I'll leave it for a potential follow up PR and remove the config_mode creating two weather entities with hourly and daily forecast.

MartinHjelmare replied:
Yeah, remove the config flow option and have one entity enabled by default and the other entity disabled by default.

Noltari replied:
@MartinHjelmare done in https://github.com/home-assistant/core/pull/45074/commits/53af241cc5b9b553758b0afdc30646feca831f4c

MatthewFlamm replied:
NWS has a similar, although it seems not as convoluted scheme for the data.  There is 1 coordinator for the current observations, and 1 coordinator for each forecast type.  Each entity gets passed 2 coordinators, the current observation and the relevant forecast coordinator.  This only makes sense if you can use separate the logic into separate coordinators.  If not, then this is a good reason to reject this structure.

I agree that given the complexity this should be a follow-on PR, if possible, to make the integration even better.

MartinHjelmare commented on homeassistant/components/aemet/weather_update_coordinator.py:
```diff
@@ -0,0 +1,652 @@
+                station_dt = dt_util.parse_datetime(
+                    station_data[AEMET_ATTR_STATION_DATE] + "Z"
+                )
+                station_timestamp = dt_util.as_utc(station_dt)
```
Sensors with device class timestamp should return a state value with an absolute UTC datetime isoformatted string. We seem to be missing the isoformatting?

Noltari replied:
Should I call utcfromtimestamp then?

MatthewFlamm reviewed (commented):
I did not review the sensor platform, only the weather platform.

MartinHjelmare commented on homeassistant/components/aemet/translations/en.json:
```diff
@@ -0,0 +1,32 @@
+{
```
We allow committing the default English translation to make it easier to test during development.

Noltari replied:
I need them for testing purposes, but I can remove them if needed.
BTW, if they are submitted, they will just be replaced, right?

Noltari commented on homeassistant/components/aemet/__init__.py:
```diff
@@ -0,0 +1,97 @@
+    """Set up AEMET OpenData as config entry."""
+    name = config_entry.data[CONF_NAME]
+    api_key = config_entry.data[CONF_API_KEY]
+    latitude = config_entry.data.get(CONF_LATITUDE, hass.config.latitude)
```
I based my integration on OpenWeatherMap, so this is inherited from there, like most of the code:
https://github.com/home-assistant/core/blob/dev/homeassistant/components/openweathermap/__init__.py#L45

MatthewFlamm replied:
It looks like this piece in openweathermap was due to legacy config in yaml and is probably no longer needed.  See https://github.com/home-assistant/core/pull/39839#issuecomment-720289686

Noltari commented on homeassistant/components/aemet/__init__.py:
```diff
@@ -0,0 +1,97 @@
+    api_key = config_entry.data[CONF_API_KEY]
+    latitude = config_entry.data.get(CONF_LATITUDE, hass.config.latitude)
+    longitude = config_entry.data.get(CONF_LONGITUDE, hass.config.longitude)
+    forecast_mode = _get_config_value(config_entry, CONF_MODE)
```
Do you mean that I should modify it to have two weather entities? (one for hourly and another one for daily)
Actually, I was planing to add another forecast mode (both?) in further PRs, after this initial one was accepted.

MatthewFlamm replied:
Yes, expose all entities possible per configuration unless there are api call concerns.  You can use this to control which are enabled by deafult. https://developers.home-assistant.io/docs/entity_registry_disabled_by/#integrations-setting-default-value-of-disabled_by-for-new-entity-registry-entries

Noltari replied:
@MatthewFlamm @MartinHjelmare I removed CONFIG_MODE and created two weather entities (and their corresponding sensors) in https://github.com/home-assistant/core/pull/45074/commits/c061576b90caf7b09693cc36261b29d3473db43a.

Noltari commented on homeassistant/components/aemet/config_flow.py:
```diff
@@ -0,0 +1,104 @@
+                vol.Required(CONF_API_KEY): str,
+                vol.Optional(CONF_NAME, default=DEFAULT_NAME): str,
+                vol.Optional(
+                    CONF_LATITUDE, default=self.hass.config.latitude
```
As above, this is inherited from OpenWeatherMap: https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/homeassistant/components/openweathermap/config_flow.py#L73

Noltari commented on homeassistant/components/aemet/config_flow.py:
```diff
@@ -0,0 +1,104 @@
+                vol.Optional(
+                    CONF_LONGITUDE, default=self.hass.config.longitude
+                ): cv.longitude,
+                vol.Optional(CONF_MODE, default=DEFAULT_FORECAST_MODE): vol.In(
```
As above, this is inherited from OpenWeatherMap: https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/homeassistant/components/openweathermap/config_flow.py#L78

Noltari commented on homeassistant/components/aemet/translations/es.json:
```diff
@@ -0,0 +1,32 @@
+{
```
Same as above.

Noltari commented:
@MatthewFlamm @MartinHjelmare honestly, I decided to start from an existing integration (OpenWeatherMap) precisely to avoid this kind of change requests.
I will try to fix them, but this looks a little unfair... (or at least from my point of view)

MatthewFlamm commented:
Sorry to hear that @Noltari .  I know that the PR process here can be lengthy, and for that reason disheartening, but I suggest you stick with it as you are close.

It looks like there are several places in which openweathermap can be improved based on this.  In addition, openweathermap has a limit of api calls per day, which makes it critical there to control what types of calls are being made in configuration.  Most of the weather api's from government sources do not have strict API call limits (although they may restrict based off of extreme usage).

Noltari commented:
> Sorry to hear that @Noltari .  I know that the PR process here can be lengthy, and for that reason disheartening, but I suggest you stick with it as you are close.

@MatthewFlamm believe me, I'm used to upstreaming stuff to other projects and I know how lengthy and disheartening it can be, but only when it makes sense.
However, requesting this amount of changes on my integration when there's an existing integration with that same code and it's not a problem for it, it's simply unfair.
Stuff like this is what discourages people from upstreaming their work to projects.
I'm seriously considering to go back to an user level and waiting for someone else to fix the problems or submit new integrations.

MartinHjelmare commented:
There's nothing unfair in our comments here. Generally it's not possible to just assume that an existing integration has no mistakes or keeps our current standards. Many of our integrations are old and often have legacy code.

Most comments here are also not hard to resolve. Regarding the possible refactor into multiple coordinators I think that's optional for this PR to avoid increasing the scope. It can be an improvement in a future PR.

Noltari commented:
> There's nothing unfair in our comments here. Generally it's not possible to just assume that an existing integration has no mistakes or keeps our current standards. Many of our integrations are old and often have legacy code.
> 
> 
> 
> Most comments here are also not hard to resolve. Regarding the possible refactor into multiple coordinators I think that's optional for this PR to avoid increasing the scope. It can be an improvement in a future PR.

It may not seem unfair to you, but it is to external developers. Documentation on how to properly create integrations is never up to date, and if non home-assistant developers can't rely on existing integrations, what can they do then? 
They just have to deal with never ending reviews to get into stale PRs which seem that are never going to get merged.

I'm not talking about how hard the comments are to resolve. I'm talking about making it easy for external developers to contribute to the project, thus resulting in project improvements.

I really think that you guys should reconsider this, because this is a real issue. The process for accepting contributions can't be this tedious and exhausting or the people will get tired of the situation and stop contributing.

MartinHjelmare commented:
We are always open for for PRs to improve our documentation. The state of our documentation is not an argument that we should accept code that is not in line with our design architecture, bad patterns or duplicate code etc.

Review and addressing comments during review is standard procedure. There's nothing out of the ordinary in this PR review.

That it takes a long time to get a review is just a function of the number of incoming PRs divided by the number of people doing reviews. The only way to improve that is to get more people doing review and we encourage that.

I'm not going to discuss our process more in this PR.

MatthewFlamm commented:
As an outside developer myself, I find the developer documentation to be very good.  And when I've had questions, I've found others willing to help point to the right direction, and then I've added to the developer documentation to clarify.  It is only as good as the developers who engage positively.

Noltari committed:
aemet: remove duplicated default domain

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: use a list comprehension instead of filter + lambda

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: abstract_aemet_sensor: inherit CoordinatorEntity

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: sensor: state: remove get() default value

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: sensor: state: simplify forecasts check

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: weather_update_coordinator: start constructor with hass

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: weather_update_coordinator: only wrap the line that can raise exceptions

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: weather_update_coordinator: isoformat station timestamps

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: init: remove default latitude/longitude

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: remove spanish translation

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: weather: inherit CoordinatorEntity

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: weather_update_coordinator: make AemetWeather a dataclass

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: remove CONFIG_MODE

Two weather entities (and their corresponding sensors) will now be created:
one with daily forecast and another one with hourly forecast.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: disable hourly weather and sensors by default

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: tests: move unload to init

Also remove unneeded options test

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

MatthewFlamm commented on homeassistant/components/aemet/abstract_aemet_sensor.py:
```diff
@@ -0,0 +1,77 @@
+        return self._unique_id
+
+    @property
+    def should_poll(self):
```
This can be removed it is in `CoordinatorEntity`

MatthewFlamm commented on homeassistant/components/aemet/abstract_aemet_sensor.py:
```diff
@@ -0,0 +1,77 @@
+        return {ATTR_ATTRIBUTION: ATTRIBUTION}
+
+    @property
+    def available(self):
```
This can be removed it is in CoordinatorEntity

MatthewFlamm commented on homeassistant/components/aemet/abstract_aemet_sensor.py:
```diff
@@ -0,0 +1,77 @@
+        """Return True if entity is available."""
+        return self.coordinator.last_update_success
+
+    async def async_added_to_hass(self):
```
This can be removed it is in CoordinatorEntity

MatthewFlamm commented on homeassistant/components/aemet/abstract_aemet_sensor.py:
```diff
@@ -0,0 +1,77 @@
+            self.coordinator.async_add_listener(self.async_write_ha_state)
+        )
+
+    async def async_update(self):
```
This can be removed it is in CoordinatorEntity

MatthewFlamm commented on homeassistant/components/aemet/weather.py:
```diff
@@ -0,0 +1,134 @@
+        return self.coordinator.data[ATTR_API_PRESSURE]
+
+    @property
+    def should_poll(self):
```
This can be removed it is in CoordinatorEntity

MatthewFlamm commented on homeassistant/components/aemet/weather.py:
```diff
@@ -0,0 +1,134 @@
+        """Return the temperature."""
+        return self.coordinator.data[ATTR_API_WIND_SPEED]
+
+    async def async_added_to_hass(self):
```
This can be removed it is in CoordinatorEntity

MatthewFlamm commented on homeassistant/components/aemet/weather.py:
```diff
@@ -0,0 +1,134 @@
+            self.coordinator.async_add_listener(self.async_write_ha_state)
+        )
+
+    async def async_update(self):
```
This can be removed it is in CoordinatorEntity

Noltari committed:
aemet: remove unneeded CoordinatorEntity functions

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

MatthewFlamm commented on tests/components/aemet/util.py:
```diff
@@ -0,0 +1,101 @@
+                CONF_API_KEY: "mock",
+                CONF_LATITUDE: "40.30403754",
+                CONF_LONGITUDE: "-3.72935236",
+                CONF_MODE: FORECAST_MODE_DAILY,
```
This is no longer used?

MatthewFlamm commented on tests/components/aemet/util.py:
```diff
@@ -0,0 +1,101 @@
+    CONF_API_KEY,
+    CONF_LATITUDE,
+    CONF_LONGITUDE,
+    CONF_MODE,
```
if not used, remove here too

MatthewFlamm commented on homeassistant/components/aemet/strings.json:
```diff
@@ -0,0 +1,32 @@
+                    "api_key": "[%key:common::config_flow::data::api_key%]",
+                    "latitude": "[%key:common::config_flow::data::latitude%]",
+                    "longitude": "[%key:common::config_flow::data::longitude%]",
+                    "mode": "[%key:common::config_flow::data::mode%]",
```
this is no longer used?

MatthewFlamm commented on homeassistant/components/aemet/strings.json:
```diff
@@ -0,0 +1,32 @@
+            }
+        }
+    },
+    "options": {
```
This whole section too, is no longer used?  I assume you would need to clean up the translations provided in the PR, but I'm not really sure on this point.

MatthewFlamm commented:
This is looking good on my re-read after the changes

Noltari committed:
aemet: completely remove CONF_MODE

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

MartinHjelmare commented on homeassistant/components/aemet/sensor.py:
```diff
@@ -0,0 +1,114 @@
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._weather_coordinator.data.get(self._sensor_type, None)
```
```suggestion
        return self._weather_coordinator.data.get(self._sensor_type)
```

MartinHjelmare commented on homeassistant/components/aemet/__init__.py:
```diff
@@ -0,0 +1,87 @@
+            hass.config_entries.async_forward_entry_setup(config_entry, component)
+        )
+
+    update_listener = config_entry.add_update_listener(async_update_options)
+    hass.data[DOMAIN][config_entry.entry_id][UPDATE_LISTENER] = update_listener
+
```
We can remove this.
```suggestion
```

MartinHjelmare commented on homeassistant/components/aemet/__init__.py:
```diff
@@ -0,0 +1,87 @@
+    return True
+
+
+async def async_update_options(hass: HomeAssistant, config_entry: ConfigEntry):
+    """Update options."""
+    await hass.config_entries.async_reload(config_entry.entry_id)
+
+
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/aemet/__init__.py:
```diff
@@ -0,0 +1,87 @@
+        )
+    )
+    if unload_ok:
+        update_listener = hass.data[DOMAIN][config_entry.entry_id][UPDATE_LISTENER]
+        update_listener()
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/aemet/config_flow.py:
```diff
@@ -0,0 +1,58 @@
+            longitude = user_input[CONF_LONGITUDE]
+
+            await self.async_set_unique_id(f"{latitude}-{longitude}")
+            self._abort_if_unique_id_configured()
```
Please test the abort case when a config entry with the same unique id as is already added to Home Assistant. Use a MockConfigEntry.

MartinHjelmare commented on tests/components/aemet/test_config_flow.py:
```diff
@@ -0,0 +1,70 @@
+        assert result["step_id"] == SOURCE_USER
+        assert result["errors"] == {}
+
+        result = await hass.config_entries.flow.async_init(
```
We normally patch the integration set up function on all create entry results to avoid setting up the whole integration. See the scaffold script tests template for example.

MatthewFlamm replied:
I don't believe this is resolved, I think Martin is asking for the setup and setup_entry functions to be patched, see https://github.com/home-assistant/core/blob/061d9c5293d1e049ee3d407aafa718891822f2dc/script/scaffold/templates/config_flow/tests/test_config_flow.py#L21_L26

Noltari replied:
@MatthewFlamm yeah, I wasn't sure about what he wanted...

Noltari replied:
> We normally patch the integration set up function on all create entry results to avoid setting up the whole integration. See the scaffold script tests template for example.

@MartinHjelmare Is this what you wanted? https://github.com/home-assistant/core/pull/45074/commits/65861c12a20c45939d97f96ad0c0caac0b87166c

Noltari committed:
aemet: remove unneeded code

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: test: config_flow: check duplicated unique id

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
tests: config_flow: api offline: use MagicMock

Use MagicMock instead of setting up the whole integration.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
tests: aemet: config_flow: test_form: mock setup_entry functions

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: remove unneeded code

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented:
Added https://github.com/home-assistant/core/pull/45074/commits/a0b177b36e919cedc06d671de5d1d9e4714c5fff removing more unneeded code and rebased to avoid homekit test errors.

Noltari committed:
aemet: simplify AemetWeather constructor

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented on tests/components/aemet/test_config_flow.py:
```diff
@@ -0,0 +1,108 @@
+        entry = MockConfigEntry(domain=DOMAIN, unique_id="aemet_unique_id", data=CONFIG)
+        entry.add_to_hass(hass)
+
+        assert await hass.config_entries.async_setup(entry.entry_id)
+        await hass.async_block_till_done()
+        assert entry.state == ENTRY_STATE_LOADED
+
+        entry_dup = MockConfigEntry(
+            domain=DOMAIN, unique_id="aemet_unique_id", data=CONFIG
+        )
+        entry_dup.add_to_hass(hass)
+
+        await hass.config_entries.async_unload(entry_dup.entry_id)
+        await hass.async_block_till_done()
+        assert entry_dup.state == ENTRY_STATE_NOT_LOADED
```
```suggestion
        result = await hass.config_entries.flow.async_init(
            DOMAIN, context={"source": SOURCE_USER}, data=CONFIG
        )
        
        assert result["type"] == "abort"
        assert result["reason" == "already_configured"
```

MartinHjelmare commented on tests/components/aemet/test_config_flow.py:
```diff
@@ -0,0 +1,108 @@
+        assert result["step_id"] == SOURCE_USER
+        assert result["errors"] == {}
+
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_USER}, data=CONFIG
+        )
```
```suggestion
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"], CONFIG
        )
```

MartinHjelmare reviewed (changes requested):
The tests needs some adjusting.

MartinHjelmare committed:
Fix tests

MartinHjelmare reviewed (approved):
Can be merged when build passes.

MartinHjelmare commented on tests/components/aemet/test_config_flow.py:
```diff
@@ -0,0 +1,102 @@
+        )
+        
+        assert result["type"] == "abort"
+        assert result["reason" == "already_configured"
```
```suggestion
        assert result["reason"] == "already_configured"
```

MartinHjelmare committed:
Fix missing bracket

Noltari committed:
aemet: fix tests

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

MartinHjelmare commented on tests/components/aemet/test_config_flow.py:
```diff
@@ -0,0 +1,87 @@
+        assert len(mock_setup.mock_calls) == 1
+        assert len(mock_setup_entry.mock_calls) == 1
+
+        entry = MockConfigEntry(domain=DOMAIN, unique_id="aemet_unique_id", data=CONFIG)
```
If we're going to test the abort case in this test too, we don't need the MockConfigEntry.

MartinHjelmare replied:
This isn't using the correct unique_id either.

Noltari replied:
Then I will revert my commit and go back to the separate tests, but I don't really know what was wrong after your changes @MartinHjelmare

MartinHjelmare replied:
The unique id isn't correct I think. It should be the mocked coordinates.

Noltari committed:
Revert "aemet: fix tests"

This reverts commit bf7d4665b812a2f92910f54a37ede006ccb10bf6.

Noltari committed:
aemet: fix tests

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

MartinHjelmare commented on tests/components/aemet/test_config_flow.py:
```diff
@@ -59,13 +60,31 @@ async def test_form(hass):
             DOMAIN, context={"source": SOURCE_USER}, data=CONFIG
         )
 
+        assert result["type"] == "create_entry"
```
Remove this and set the correct unique_id in the MockConfigEntry instead.

Noltari committed:
tests: aemet: test_form_duplicated_id: properly mock config entry

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

MartinHjelmare reviewed (approved):
Thanks!

Noltari committed:
tests: aemet: fix black

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>
